### PR TITLE
Avoid frequent dynamic_cast to check if a node is an intermediate.

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe or @arnej27959 PR
I will remove the cast one at a time as I suspect the remaining one is more expensive than the other.

### CODE REVIEW

**Review by arnej27959 — APPROVED** [review]
👍

File: searchcore/src/vespa/searchcore/proton/matching/query.cpp [hunk 1]
```diff
@@ -64,11 +64,14 @@ find_location_terms(Node *tree) {
     std::vector<ProtonLocationTerm *> retval;
     std::vector<Node *> nodes;
     nodes.push_back(tree);
+    // Note the nodes vector being iterated over is appended in the loop
     for (size_t i = 0; i < nodes.size(); ++i) {
-        if (auto loc = dynamic_cast<ProtonLocationTerm *>(nodes[i])) {
+        Node * node = nodes[i];
+        if (auto loc = dynamic_cast<ProtonLocationTerm *>(node)) {
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/query.cpp:70]: if you are looking for only a single specific class (ensure no sub-classes by making it final), using std::type_index(typeid(...)) is much faster than dynamic cast (and even faster than virtual predicates combined with static cast)

File: searchcore/src/vespa/searchcore/proton/matching/query.cpp [hunk 2]
```diff
@@ -64,11 +64,14 @@ find_location_terms(Node *tree) {
...
             retval.push_back(loc);
         }
-        if (auto parent = dynamic_cast<const search::query::Intermediate *>(nodes[i])) {
+        if (node->isIntermediate()) {
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/query.cpp:73]: a node cannot be both intermediate and a location term at a same time, consider using 'else' to model exclusivity.

Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/query.cpp:73]: That is fixed in an upcoming PR

File: searchcore/src/vespa/searchcore/proton/matching/query.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/query.cpp:70]: hmm, that one is new to me. Will take a look.

Comment (havardpe) [reply searchcore/src/vespa/searchcore/proton/matching/query.cpp:70]: various strategies are tested in 'vespa/vespalib/src/tests/detect_type_benchmark'

**Pull request merged**